compiler/optimizations/gpuTransforms.cpp

```diff
@@ -169,6 +171,8 @@ class GpuizableLoop {
   std::vector<Symbol*> lowerBounds_;
   bool shouldErrorIfNotGpuizable_;
 
+  std::vector<CallExpr*> gpuAssertions_; // should this be a single CallExpr* ?
```

Comment:
Good point. To be clear, in that code, you'd want to get a warning from both of these assertions, right? Do you also want that code to be valid if the `proc` wasn't `inline`?

I am not entirely sure if there's an easy way to emit that warning only once per loop, especially for the case where the called function is not inline. Assertions in the loop body are somewhat easier as you can replace them with:


```chapel
var asserted = false;
if !asserted {
  chpl_gpu_assert_on_gpu();
  asserted = true;
}
```

That's not so easy to do for a function that's not inlined, though.

With your general comment about relocating assertions, would you think asserting per iteration but using a `config const` to suppress that warning a better approach?

Reply:
> Good point. To be clear, in that code, you'd want to get a warning from both of these assertions, right? Do you also want that code to be valid if the proc wasn't inline?

I'm not sure, I'm starting to have second thoughts if its even worth having it warn it all. It seems like it would be a real nuisance to get the warning multiple times (especially if it becomes once per thread). So I think if we do want to keep `assertOnGpu` and we want it to warn when "passing" with `CHPL_GPU=cpu` then I'd yeah I'd lean towards having a highly advertised `config const` to suppress it.

I still think we might want to consider something like this though:

- Having an `@assertGpuEligible` or if we want to be bold `@assertGpuizable` annotation if the user cares about having a compile-time check.
- Just telling the user to do `assert(here.isGpu())` if they want a runtime check. This would require us to get halting to work for GPUs (or maybe we could just pattern match and remove instances of it). Note that this would just silently pass if `CHPL_GPU=cpu` and the user is on one of the "simulated" gpu sublocales.

Reply:
> So I think if we do want to keep assertOnGpu and we want it to warn when "passing" with CHPL_GPU=cpu then I'd yeah I'd lean towards having a highly advertised config const to suppress it.

I think this is the safest incremental step for this PR.

> I still think we might want to consider something like this though:

I can get behind that proposal. Arguably, this is one facet of replacing the `GPU` modules features with existing Chapel features as much as possible.